### ISSUE TITLE
The plugin fails to recognise valid Ghidra installations

### ISSUE THREAD

**Issue by agatti** [garyttierney/intellij-ghidra#9]
**Describe the bug:**

The plugin reports an exception at startup about missing paths, then fails to launch Ghidra.

**Steps to reproduce:** 

* Build Ghidra from git using `gradle buildGhidra` (I'm testing with revision  42ca00b35d14a08f3105756869aa8478a591fa04)
* Unzip the archive found under `build/dist` to any directory
* Run `./ghidraRun` in the directory chosen early, then once Ghidra is started just close it
* Import the Ghidra project in IntelliJ, then add the Ghidra facet using the path from earlier
* The plugin reports an exception about an invalid path (see stack trace)

**Expected behavior:**

The plugin does not report any exception on valid Ghidra installations.

**Additional context:**

IDE: IntelliJ Ultimate 2021.3
OS: macOS 12.0.1
JDK:

openjdk version "11.0.13" 2021-10-19
OpenJDK Runtime Environment Temurin-11.0.13+8 (build 11.0.13+8)
OpenJDK 64-Bit Server VM Temurin-11.0.13+8 (build 11.0.13+8, mixed mode)

Stack trace when choosing the path (the failure occurs when enumerating `Ghidra/Extensions`, which happens to be empty:

```
java.lang.IllegalStateException: Installation path doesn't exist
	at com.codingmates.ghidra.intellij.ide.facet.GhidraFacetConfiguration.searchGhidraRoots(GhidraFacetConfiguration.kt:25)
	at com.codingmates.ghidra.intellij.ide.facet.GhidraFacetConfiguration.loadGhidraInstallation(GhidraFacetConfiguration.kt:62)
	at com.codingmates.ghidra.intellij.ide.facet.GhidraFacet$updateLibrary$$inlined$runWriteAction$1.compute(actions.kt:64)
	at com.intellij.openapi.application.impl.ApplicationImpl.lambda$runWriteAction$15(ApplicationImpl.java:956)
	at com.intellij.openapi.application.impl.ApplicationImpl.runWriteActionWithClass(ApplicationImpl.java:935)
	at com.intellij.openapi.application.impl.ApplicationImpl.runWriteAction(ApplicationImpl.java:956)
	at com.codingmates.ghidra.intellij.ide.facet.GhidraFacet.updateLibrary(GhidraFacet.kt:128)
	at com.codingmates.ghidra.intellij.ide.facet.GhidraFacet$1.facetConfigurationChanged(GhidraFacet.kt:36)
	at com.intellij.util.messages.impl.MessageBusImpl.invokeMethod(MessageBusImpl.java:645)
	at com.intellij.util.messages.impl.MessageBusImpl.invokeListener(MessageBusImpl.java:620)
	at com.intellij.util.messages.impl.MessageBusImpl.deliverMessage(MessageBusImpl.java:417)
	at com.intellij.util.messages.impl.MessageBusImpl.pumpWaitingBuses(MessageBusImpl.java:390)
	at com.intellij.util.messages.impl.MessageBusImpl.pumpMessages(MessageBusImpl.java:372)
	at com.intellij.util.messages.impl.MessageBusImpl.access$200(MessageBusImpl.java:33)
	at com.intellij.util.messages.impl.MessageBusImpl$MessagePublisher.invoke(MessageBusImpl.java:179)
	at com.sun.proxy.$Proxy216.facetConfigurationChanged(Unknown Source)
	at com.intellij.facet.impl.FacetEventsPublisher.fireFacetConfigurationChanged(FacetEventsPublisher.kt:83)
	at com.intellij.facet.FacetManagerBase.facetConfigurationChanged(FacetManagerBase.java:76)
	at com.intellij.workspaceModel.ide.impl.legacyBridge.facet.FacetManagerBridge.facetConfigurationChanged(FacetManagerBridge.kt:51)
	at com.intellij.facet.impl.ProjectFacetsConfigurator.commitFacets(ProjectFacetsConfigurator.java:235)
	at com.intellij.openapi.roots.ui.configuration.ModulesConfigurator.lambda$apply$3(ModulesConfigurator.java:348)
	at com.intellij.openapi.application.impl.ApplicationImpl.runWriteAction(ApplicationImpl.java:947)
	at com.intellij.openapi.roots.ui.configuration.ModulesConfigurator.apply(ModulesConfigurator.java:316)
	at com.intellij.openapi.roots.ui.configuration.projectRoot.ModuleStructureConfigurable.apply(ModuleStructureConfigurable.java:378)
	at com.intellij.openapi.roots.ui.configuration.ProjectStructureConfigurable.apply(ProjectStructureConfigurable.java:312)
	at com.intellij.openapi.options.ex.SingleConfigurableEditor$ApplyAction.actionPerformed(SingleConfigurableEditor.java:209)
	at java.desktop/javax.swing.AbstractButton.fireActionPerformed(AbstractButton.java:1967)
	at java.desktop/javax.swing.AbstractButton$Handler.actionPerformed(AbstractButton.java:2308)
	at java.desktop/javax.swing.DefaultButtonModel.fireActionPerformed(DefaultButtonModel.java:405)
	at java.desktop/javax.swing.DefaultButtonModel.setPressed(DefaultButtonModel.java:262)
	at java.desktop/javax.swing.plaf.basic.BasicButtonListener.mouseReleased(BasicButtonListener.java:270)
	at java.desktop/java.awt.Component.processMouseEvent(Component.java:6654)
	at java.desktop/javax.swing.JComponent.processMouseEvent(JComponent.java:3345)
	at java.desktop/java.awt.Component.processEvent(Component.java:6419)
	at java.desktop/java.awt.Container.processEvent(Container.java:2263)
	at java.desktop/java.awt.Component.dispatchEventImpl(Component.java:5029)
	at java.desktop/java.awt.Container.dispatchEventImpl(Container.java:2321)
	at java.desktop/java.awt.Component.dispatchEvent(Component.java:4861)
	at java.desktop/java.awt.LightweightDispatcher.retargetMouseEvent(Container.java:4918)
	at java.desktop/java.awt.LightweightDispatcher.processMouseEvent(Container.java:4547)
	at java.desktop/java.awt.LightweightDispatcher.dispatchEvent(Container.java:4488)
	at java.desktop/java.awt.Container.dispatchEventImpl(Container.java:2307)
	at java.desktop/java.awt.Window.dispatchEventImpl(Window.java:2790)
	at java.desktop/java.awt.Component.dispatchEvent(Component.java:4861)
	at java.desktop/java.awt.EventQueue.dispatchEventImpl(EventQueue.java:778)
	at java.desktop/java.awt.EventQueue$4.run(EventQueue.java:727)
	at java.desktop/java.awt.EventQueue$4.run(EventQueue.java:721)
	at java.base/java.security.AccessController.doPrivileged(Native Method)
	at java.base/java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:85)
	at java.base/java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:95)
	at java.desktop/java.awt.EventQueue$5.run(EventQueue.java:751)
	at java.desktop/java.awt.EventQueue$5.run(EventQueue.java:749)
	at java.base/java.security.AccessController.doPrivileged(Native Method)
	at java.base/java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:85)
	at java.desktop/java.awt.EventQueue.dispatchEvent(EventQueue.java:748)
	at com.intellij.ide.IdeEventQueue.defaultDispatchEvent(IdeEventQueue.java:891)
	at com.intellij.ide.IdeEventQueue.dispatchMouseEvent(IdeEventQueue.java:820)
	at com.intellij.ide.IdeEventQueue._dispatchEvent(IdeEventQueue.java:757)
	at com.intellij.ide.IdeEventQueue.lambda$dispatchEvent$6(IdeEventQueue.java:447)
	at com.intellij.openapi.progress.impl.CoreProgressManager.computePrioritized(CoreProgressManager.java:818)
	at com.intellij.ide.IdeEventQueue.lambda$dispatchEvent$7(IdeEventQueue.java:446)
	at com.intellij.openapi.application.impl.ApplicationImpl.runIntendedWriteActionOnCurrentThread(ApplicationImpl.java:805)
	at com.intellij.ide.IdeEventQueue.dispatchEvent(IdeEventQueue.java:498)
	at java.desktop/java.awt.EventDispatchThread.pumpOneEventForFilters(EventDispatchThread.java:203)
	at java.desktop/java.awt.EventDispatchThread.pumpEventsForFilter(EventDispatchThread.java:124)
	at java.desktop/java.awt.EventDispatchThread.pumpEventsForFilter(EventDispatchThread.java:117)
	at java.desktop/java.awt.WaitDispatchSupport$2.run(WaitDispatchSupport.java:190)
	at java.desktop/java.awt.WaitDispatchSupport$4.run(WaitDispatchSupport.java:235)
	at java.desktop/java.awt.WaitDispatchSupport$4.run(WaitDispatchSupport.java:233)
	at java.base/java.security.AccessController.doPrivileged(Native Method)
	at java.desktop/java.awt.WaitDispatchSupport.enter(WaitDispatchSupport.java:233)
	at java.desktop/java.awt.Dialog.show(Dialog.java:1070)
	at com.intellij.openapi.ui.impl.DialogWrapperPeerImpl$MyDialog.show(DialogWrapperPeerImpl.java:701)
	at com.intellij.openapi.ui.impl.DialogWrapperPeerImpl.show(DialogWrapperPeerImpl.java:437)
	at com.intellij.openapi.ui.DialogWrapper.doShow(DialogWrapper.java:1671)
	at com.intellij.openapi.ui.DialogWrapper.show(DialogWrapper.java:1629)
	at com.intellij.ide.actions.ShowStructureSettingsAction.showDialog(ShowStructureSettingsAction.java:46)
	at com.intellij.ide.actions.ShowStructureSettingsAction.actionPerformed(ShowStructureSettingsAction.java:36)
	at com.intellij.openapi.keymap.impl.ActionProcessor.performAction(ActionProcessor.java:65)
	at com.intellij.openapi.keymap.impl.IdeKeyEventDispatcher$1.performAction(IdeKeyEventDispatcher.java:572)
	at com.intellij.openapi.keymap.impl.IdeKeyEventDispatcher.lambda$doPerformActionInner$10(IdeKeyEventDispatcher.java:702)
	at com.intellij.openapi.actionSystem.ex.ActionUtil.performDumbAwareWithCallbacks(ActionUtil.java:265)
	at com.intellij.openapi.keymap.impl.IdeKeyEventDispatcher.doPerformActionInner(IdeKeyEventDispatcher.java:698)
	at com.intellij.openapi.keymap.impl.IdeKeyEventDispatcher.processAction(IdeKeyEventDispatcher.java:641)
	at com.intellij.openapi.keymap.impl.IdeKeyEventDispatcher.processAction(IdeKeyEventDispatcher.java:583)
	at com.intellij.openapi.keymap.impl.IdeKeyEventDispatcher.processActionOrWaitSecondStroke(IdeKeyEventDispatcher.java:466)
	at com.intellij.openapi.keymap.impl.IdeKeyEventDispatcher.inInitState(IdeKeyEventDispatcher.java:455)
	at com.intellij.openapi.keymap.impl.IdeKeyEventDispatcher.dispatchKeyEvent(IdeKeyEventDispatcher.java:223)
	at com.intellij.ide.IdeEventQueue.dispatchKeyEvent(IdeEventQueue.java:804)
	at com.intellij.ide.IdeEventQueue._dispatchEvent(IdeEventQueue.java:754)
	at com.intellij.ide.IdeEventQueue.lambda$dispatchEvent$6(IdeEventQueue.java:447)
	at com.intellij.openapi.progress.impl.CoreProgressManager.computePrioritized(CoreProgressManager.java:818)
	at com.intellij.ide.IdeEventQueue.lambda$dispatchEvent$7(IdeEventQueue.java:446)
	at com.intellij.openapi.application.impl.ApplicationImpl.runIntendedWriteActionOnCurrentThread(ApplicationImpl.java:805)
	at com.intellij.ide.IdeEventQueue.dispatchEvent(IdeEventQueue.java:498)
	at java.desktop/java.awt.EventDispatchThread.pumpOneEventForFilters(EventDispatchThread.java:203)
	at java.desktop/java.awt.EventDispatchThread.pumpEventsForFilter(EventDispatchThread.java:124)
	at java.desktop/java.awt.EventDispatchThread.pumpEventsForHierarchy(EventDispatchThread.java:113)
	at java.desktop/java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:109)
	at java.desktop/java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:101)
	at java.desktop/java.awt.EventDispatchThread.run(EventDispatchThread.java:90)```

**Comment by sud0why** [comment]
You can debug that problem. I locate this problem at `main/src/main/kotlin/com/codingmates/ghidra/intellij/ide/facet/GhidraFacetConfiguration.kt`, line 25, I edit this file to 

```
val vfs = VirtualFileManager.getInstance()
val strpath = root.toString()
val vfsRoot = vfs.findFileByNioPath(root) ?: error("Installation path $strpath doesn't exist")
```

 when it exception , i  can see this plugin want `C:\Users\user\.ghidra\.ghidra_10.1_PUBLIC\Extensions` dir, but `Extensions` dir is not exist. So I create this dir manualy. This plugin work!

**Comment by agatti** [comment]
@sud0why probably you missed the bit where I mentioned that `Extensions` exists but it is empty.  The issue is that the plugin errors out if there is *any* empty directory somewhere within Ghidra when it is installed using the procedure mentioned earlier.

**Comment by garyttierney** [comment]
> The issue is that the plugin errors out if there is any empty directory somewhere within Ghidra when it is installed using the procedure mentioned earlier.

Huh, that is quite bizarre. I can take a look at this today.

**Comment by agatti** [comment]
Would something like this be ok then?

```diff

--- a/src/main/kotlin/com/codingmates/ghidra/intellij/ide/facet/GhidraFacetConfiguration.kt
+++ b/src/main/kotlin/com/codingmates/ghidra/intellij/ide/facet/GhidraFacetConfiguration.kt
@@ -22,16 +22,16 @@ class GhidraFacetConfiguration : FacetConfiguration, PersistentStateComponent<Gh
     private fun searchGhidraRoots(root: Path, filter: (VirtualFile) -> Boolean): MutableList<VirtualFile> {
         val roots = mutableListOf<VirtualFile>()
         val vfs = VirtualFileManager.getInstance()
-        val vfsRoot = vfs.findFileByNioPath(root) ?: error("Installation path doesn't exist")
-
-        VfsUtil.iterateChildrenRecursively(vfsRoot, { !it.path.contains("GhidraServer/data") }) {
-            if (filter(it)) {
-                val uri = VfsUtil.getUrlForLibraryRoot(it.toNioPath().toFile())
-                val archiveFile = vfs.findFileByUrl(uri) ?: return@iterateChildrenRecursively true
-                roots.add(archiveFile)
+        vfs.findFileByNioPath(root)?.let { vfsRoot ->
+            VfsUtil.iterateChildrenRecursively(vfsRoot, { !it.path.contains("GhidraServer/data") }) {
+                if (filter(it)) {
+                    val uri = VfsUtil.getUrlForLibraryRoot(it.toNioPath().toFile())
+                    val archiveFile = vfs.findFileByUrl(uri) ?: return@iterateChildrenRecursively true
+                    roots.add(archiveFile)
+                }
+
+                true
             }
-
-            true
         }

         return roots
```

**Comment by garyttierney** [comment]
Ah, I see. We explicitly attempt to search the Extension directory in `loadGhidraInstallation()`. Couldn't quite recall why `searchGhidraRoots` would be called with the extension folder as the root.

Yep, that patch LGTM. I'll commit that after work or if you want to open a PR I'm happy to merge it in.

**Comment by garyttierney** [comment]
This should be resolved now that your patch is in master. I'll close once I cut a new release and have binaries out with the fix.

**Comment by agatti** [comment]
Awesome, thanks!

**Comment by garyttierney** [comment]
Released in v0.3.0. Should hopefully be the first version on the plugin marketplace too.
